components/raftstore/src/store/worker/cleanup_snapshot.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2018 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{fmt, marker::PhantomData, ops::Deref};
+
+use crossbeam::channel::TrySendError;
+use engine_traits::{KvEngine, RaftEngine};
+use fail::fail_point;
+use tikv_util::{debug, error, info, warn, worker::Runnable};
+
+use crate::{
+    store::{
+        CasualMessage, PeerMsg, RaftRouter, SnapKey, SnapManager, Snapshot, StoreMsg, StoreRouter,
+    },
+    Result,
+};
+
+pub enum Task {
+    GcSnapshot,
+    DeleteSnapshotFiles {
+        key: SnapKey,
+        snapshot: Box<Snapshot>,
+        check_entry: bool,
+    },
+}
+
+impl fmt::Display for Task {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &*self {
+            Task::GcSnapshot => write!(f, "Gc Snapshot"),
+            Task::DeleteSnapshotFiles { key, .. } => write!(f, "Delete Snapshot Files for {}", key),
+        }
+    }
+}
+
+pub struct Runner<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    store_id: u64,
+    router: RaftRouter<EK, ER>,
+    snap_mgr: SnapManager,
+    _engine: PhantomData<EK>,
+}
+
+impl<EK, ER> Runner<EK, ER>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+{
+    pub fn new(store_id: u64, router: RaftRouter<EK, ER>, snap_mgr: SnapManager) -> Runner<EK, ER> {
+        Runner {
+            store_id,
+            router,
+            snap_mgr,
+            _engine: PhantomData,
+        }
+    }
+
+    /// handle snap mgr gc
+    fn handle_snap_mgr_gc(&self) -> Result<()> {
+        fail_point!("peer_2_handle_snap_mgr_gc", self.store_id == 2, |_| Ok(()));
+        let snap_keys = self.snap_mgr.list_idle_snap()?;
+        if snap_keys.is_empty() {
+            return Ok(());
+        }
+        let (mut last_region_id, mut keys) = (0, vec![]);
+        let schedule_gc_snap = |region_id: u64, snaps| -> Result<()> {
+            debug!(
+                "schedule snap gc";
+                "store_id" => self.store_id,
+                "region_id" => region_id,
+            );
+
+            let gc_snap = PeerMsg::CasualMessage(CasualMessage::GcSnap { snaps });
+            match Deref::deref(&self.router).send(region_id, gc_snap) {
```

Comment:
`self.router.send()` is not workable?

Reply:
It will have following error: ---the problem is that gc_snap is PeerMsg, not CasualMessage (so I cannot use CasualRouter). So it seems I have to call its defer API.  But looks (*self.router).send works fine. 

error[E0061]: this function takes 1 argument but 2 arguments were supplied
  --> components/raftstore/src/store/worker/cleanup_snapshot.rs:76:31
   |
76 |             match self.router.send(region_id, gc_snap) {
   |                               ^^^^ ---------  ------- supplied 2 arguments
   |                               |
   |                               expected 1 argument

Reply:
You mean `CasualRouter::send(&self.router, region_id, gc_snap)`?

Reply:
> 
Not CasualRouter because the msg here is PeerMessage, not CasualMessage. I need to call BatchRouter's send, which is done by deref self.router.

Reply:
The reason that it cannot use CasualRouter::send(&self, region_id: u64, msg: CasualMessage<EK>)  is because it's error processing does not meet the need and we have to use the underlying router's send, which only accept PeerMsg as message type.